lib/ui/painting/image_decoder.cc

```diff
@@ -57,30 +62,91 @@ static SkISize GetResizedDimensions(SkISize current_size,
                                         std::optional<uint32_t> target_width,
                                         std::optional<uint32_t> target_height,
                                         const fml::tracing::TraceFlow& flow) {
   FML_DCHECK(!image->isTextureBacked());
 
-  const auto resized_dimensions =
+  TRACE_EVENT0("flutter", __FUNCTION__);
+  flow.Step(__FUNCTION__);
+
+  if (!target_width && !target_height) {
+    // The resized dimesions are the same as the intrinsic dimensions of the
+    // image. There is nothing to do.
+    return image->makeRasterImage();
+  }
+
+  auto resized_dimensions =
       GetResizedDimensions(image->dimensions(), target_width, target_height);
 
   if (resized_dimensions.isEmpty()) {
     FML_LOG(ERROR) << "Could not resize to empty dimensions.";
     return nullptr;
   }
 
-  if (resized_dimensions == image->dimensions()) {
-    // The resized dimesions are the same as the intrinsic dimensions of the
-    // image. There is nothing to do.
-    return image;
+  // This is where the scaling happens.
+  if (resized_dimensions.width() > image->dimensions().width() ||
+      resized_dimensions.height() > image->dimensions().height()) {
+    FML_LOG(WARNING)
+        << "Image is being upsized. Using cache(Height|Width) correctly?";
+    // TOOD(jeltsch): consider exiting here, there's no good reason to support
+    // upsampling in a "caching"-optimiztion context..
   }
 
-  TRACE_EVENT0("flutter", __FUNCTION__);
-  flow.Step(__FUNCTION__);
+  const bool aspect_ratio_changed =
+      std::abs(AspectRatio(resized_dimensions) -
+               AspectRatio(image->dimensions())) > 0.01;
+  if (aspect_ratio_changed) {
+    // This is probably a bug. If a user passes dimensions that change the
+    // aspect ratio in a "caching" context that's probably not working as
+    // intended and rather a signal that the API is hard to use.
+    FML_LOG(WARNING) << "Aspect ratio is being changed.";
+  }
 
-  const auto scaled_image_info = image->imageInfo().makeWH(
-      resized_dimensions.width(), resized_dimensions.height());
+  if (codec != nullptr) {
+    auto decode_size = codec->getScaledDimensions(
+        std::max(static_cast<double>(resized_dimensions.width()) /
+                     image->dimensions().width(),
+                 static_cast<double>(resized_dimensions.height()) /
+                     image->dimensions().height()));
+
+    // If the codec supports efficient sub-sampling during decode use the
+    // supported size instead.
+    // Note that this won't respect the exact user-requested size but instead
+    // choose a supported approximation.
+    if (decode_size != image->dimensions()) {
+      const auto scaled_image_info =
+          image->imageInfo().makeDimensions(decode_size);
+      SkBitmap scaled_bitmap;
+      if (!scaled_bitmap.tryAllocPixels(scaled_image_info)) {
+        FML_LOG(ERROR) << "Could not allocate bitmap when attempting to scale.";
```

Comment:
What is a developer supposed to do about this?

AFAICT this can happen if the image data is corrupted, or if there's not enough memory.  Maybe we could just add something here to that effect?

Reply:
I asked myself the same question. I'm merely preserving the prior message: https://github.com/flutter/engine/blob/5ef2c7968c7215acaffddd008a8ccabeb7cb6706/lib/ui/painting/image_decoder.cc#L87.

Would you like me to take a stab at it and update the message consistently with L87 in this change?

Reply:
I think our existing message violates our style guide :)

It's fine to log an error here, but IMO saying why would help.  `"Could not allocate bitmap when attempting to scale image. Check that the image data is valid and that there is enough memory available for allocating " << (decode_size.width() * decode_size.height() * 4) << " bytes."` or something like that.

Reply:
Updated the message to contain bytes and put an emphasis on memory.